s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java

```diff
@@ -227,31 +227,48 @@ protected void handleAttachException( String key, String bucket ) throws IOExcep
       s3ObjectMetadata = fileSystem.getS3Client().getObjectMetadata( bucketName, key );
       injectType( FileType.FOLDER );
       this.key = keyWithDelimiter;
-    } catch ( AmazonS3Exception e2 ) {
-      ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
-        .withBucketName( bucket )
-        .withPrefix( keyWithDelimiter )
-        .withDelimiter( DELIMITER );
-      ObjectListing ol = fileSystem.getS3Client().listObjects( listObjectsRequest );
-
-      if ( !( ol.getCommonPrefixes().isEmpty() && ol.getObjectSummaries().isEmpty() ) ) {
-        injectType( FileType.FOLDER );
-      } else {
-        //Folders don't really exist - they will generate a "NoSuchKey" exception
-        String errorCode = e2.getErrorCode();
-        // confirms key doesn't exist but connection okay
-        if ( !errorCode.equals( "NoSuchKey" ) ) {
-          // bubbling up other connection errors
-          logger.error( "Could not get information on " + getQualifiedName(),
-            e2 ); // make sure this gets printed for the user
-          throw new FileSystemException( "vfs.provider/get-type.error", getQualifiedName(), e2 );
+    } catch ( AmazonS3Exception e1 ) {
+      String errorCode = e1.getErrorCode();
+      try {
+        if ( errorCode.equals( "404 Not Found" ) ) {
```

Comment:
Please Add comments explaining why this is needed.

Reply:
Added a few comments. 
This approach is a bit conservative. Chris's implementation improved performance, so i kept it. But in this way of searching for the S3Object we end up with a not found exception, making it impossible to proceed in the case we are creating a new file. (this impacted the S3 File Output as described in the issue). 
I now catch and see if we have the file not found exception and try to retrieve it with the S3 getObject. (basically what was done before). If for some reason this goes wrong it fallback to the exception handling as always.